src/shims/panic.rs

```diff
@@ -0,0 +1,180 @@
+
+        // fn __rust_maybe_catch_panic(
+        //     f: fn(*mut u8),
+        //     data: *mut u8,
+        //     data_ptr: *mut usize,
+        //     vtable_ptr: *mut usize,
+        // ) -> u32
+        let f = this.read_scalar(args[0])?.not_undef()?;
+        let data = this.read_scalar(args[1])?.not_undef()?;
+        let data_ptr = this.deref_operand(args[2])?;
+        let vtable_ptr = this.deref_operand(args[3])?;
+        let f_instance = this.memory.get_fn(f)?.as_instance()?;
+        this.write_null(dest)?;
+        trace!("__rust_maybe_catch_panic: {:?}", f_instance);
+
+
+        // Now we make a function call.
+        // TODO: consider making this reusable? `InterpCx::step` does something similar
+        // for the TLS destructors, and of course `eval_main`.
+        let mir = this.load_mir(f_instance.def, None)?;
+        let ret_place =
+            MPlaceTy::dangling(this.layout_of(tcx.mk_unit())?, this).into();
+        this.push_stack_frame(
+            f_instance,
+            mir.span,
+            mir,
+            Some(ret_place),
+            // Directly return to caller.
+            StackPopCleanup::Goto { ret: Some(ret), unwind: None },
+        )?;
+
+        // In unwind mode, we tag this frame with some extra data.
+        // This lets '__rust_start_panic' know that we should stop unwinding
+        // when we pop this frame
+        if this.tcx.tcx.sess.panic_strategy() == PanicStrategy::Unwind {
+            this.frame_mut().extra.catch_panic = Some(CatchUnwindData {
+                data: data.to_ptr()?,
+                data_ptr,
+                vtable_ptr,
+                dest: dest.clone(),
+                ret,
+            })
+        }
+
+        let mut args = this.frame().body.args_iter();
+
+        let arg_local = args
+            .next()
+            .expect("Argument to __rust_maybe_catch_panic does not take enough arguments.");
+        let arg_dest = this.local_place(arg_local)?;
+        this.write_scalar(data, arg_dest)?;
+
+        args.next().expect_none("__rust_maybe_catch_panic argument has more arguments than expected");
+
+        // We ourselves will return `0`, eventually (because we will not return if we paniced).
+        this.write_null(dest)?;
+
+        return Ok(());
+    }
+
+    #[inline(always)]
+    fn handle_stack_pop(
+        &mut self,
+        mut extra: FrameData<'tcx>,
+        unwinding: bool
+    ) -> InterpResult<'tcx, StackPopInfo> {
+        let this = self.eval_context_mut();
+
+        trace!("handle_stack_pop(extra = {:?}, unwinding = {})", extra, unwinding);
+
+        // We only care about `catch_panic` if we're unwinding - if we're doing a normal
+        // return, then we don't need to do anything special.
+        let res = if let (true, Some(unwind_data)) = (unwinding, extra.catch_panic.take()) {
+            // We've just popped the frame that was immediately above
+            // our target frame on the stack.
```

Comment:
I am still not sure what you mean by "target frame", and the "immediately above" part seems outdated now.